### PR TITLE
Decouple snap from quantize

### PR DESCRIPTION
**Pull request by ninomp**
This PR seeks to decouple snapping-to-beat from quantizing.

This resolves #8970

Note: UI for enabling/disabling snapping to beat is available in LateNight skin only.

### CODE REVIEW

**Comment by daschuer** [comment]
We need at least a concept how to migrate users to the new approach. 
We need also consider if we need this approach for a all users or if this should be an option. 
I general I am in line with Dale (dj-kaza) here and like the simplicity of having just to deal with one button. 

@ninomp What was your personal motivation for this PR? Are you used to separated knobs from an other tool?

**Comment by Be-ing** [comment]
> if this should be an option

No options please.

File: src/skin/legacy/tooltips.cpp [hunk 1]
```diff
@@ -451,11 +451,15 @@ void Tooltips::addStandardTooltips() {
         <<tr("Toggle visibility of Rate Control");
 
     // Used in cue/hotcue/loop tooltips below.
-    QString quantizeSnap = tr("If quantize is enabled, snaps to the nearest beat.");
+    QString quantizeSnap = tr("If snap is enabled, snaps to the nearest beat.");
```
Comment (Be-ing) [inline src/skin/legacy/tooltips.cpp:454]:
```suggestion
    QString snapDescription = tr("If snap is enabled, snaps to the nearest beat.");
```
This variable name is confusing now.

**Comment by Holzhaus** [comment]
> > if this should be an option
> 
> No options please.

I think if we want to support this at all, then options similar to rekordbox are our best choice IMHO. I'd rather clutter the preferences than the skins.

File: src/skin/legacy/tooltips.cpp [hunk 2]
```diff
@@ -451,11 +451,15 @@ void Tooltips::addStandardTooltips() {
...
     add("quantize")
             << tr("Quantize")
-            << tr("Toggles quantization.")
-            << tr("Loops and cues snap to the nearest beat when quantization is enabled.");
+            << tr("Toggles quantization.");
```
Comment (Be-ing) [inline src/skin/legacy/tooltips.cpp:457]:
```suggestion
            << tr("Snaps playhead to the nearest beat when pressing play.");
```

File: src/skin/legacy/tooltips.cpp [hunk 3]
```diff
@@ -451,11 +451,15 @@ void Tooltips::addStandardTooltips() {
...
+
+    add("snap")
+            << tr("Snap")
+            << tr("Toggles snapping.")
+            << tr("Loops and cues snap to the nearest beat when snapping is enabled.");
```
Comment (Be-ing) [inline src/skin/legacy/tooltips.cpp:462]:
```suggestion
            << tr("Snap loop and cues to the nearest beat when placing them.");
```

**Comment by Be-ing** [comment]
Thanks for working on this! This has bugged me for years and thrown off quite a bit of my mixes.

**Comment by Be-ing** [comment]
> I think if we want to support this at all, then options similar to rekordbox are our best choice IMHO. I'd rather clutter the preferences than the skins.

I disagree. Please explain why adding a new button to skins is a problem.

**Comment by ronso0** [comment]
> > > if this should be an option
> > 
> > 
> > No options please.
> 
> I think if we want to support this at all, then options similar to rekordbox are our best choice IMHO. I'd rather clutter the preferences than the skins.

Could that also be a 4-state button:
off / snap / quantize / snap+quantize
?
challenging to design (small) icons for all states though

**Comment by Be-ing** [comment]
A 4 state button would be even worse than the current situation IMO.

**Comment by Be-ing** [comment]
Do we really need a new button for every deck? I don't really want that; I want one button that affects all decks. However for quantize we currently have a button for every deck, even though in my experience that is overcomplicated and difficult to manage in a mix.

**Comment by daschuer** [comment]
Can we step back and talk about the use cases in the bug report first? I have the impression that we can make this feature really useful for more creativity and better controller support than just copy the Traktor solution. Let's see.

**Comment by Be-ing** [comment]
Please let's move this discussion to Zulip then. Launchpad is clumsy for discussions.

**Comment by ninomp** [comment]
@daschuer My personal motivation for this PR is that I would like my cues and loops to snap to nearest beat, but I don't want track to jump when I press play. Currently, I have to enable quantize, place a cue/loop, and then disable quantize.

**Comment by ninomp** [comment]
I have come up with couple of ideas on how to fit both Snap and Quantize buttons in our skins:

1. If we are going to build new UI for editing beats of a track, we can replace the `beats_translate_curpos` button with `snap`
2. In Deere and Tango skins we could place Eject button in the row that contains track title, similar to how it is done in Shade
3. We could make the small buttons in decks be arranged in 2x4 grid instead of 2x3

**Comment by ninomp** [comment]
I will start a discussion on Zulip.

**Comment by ninomp** [comment]
This is ready for review. The only thing missing is SNAP button in Shade skin. I have no idea where to squeeze it in.

**Comment by ronso0** [comment]
Just a quick checkin (I do not really care about snap/quantize, I don't use it)

I had a brief look at the UI changes, and I'm not really happy with the SNAP button in the Mixxx 'menu row'. Where we had only UI toggles, triggers (Rec, Broadcast) and status infos, there's now also a _mode switch_.
I'm simply noting that, I can't propose a better spot for that atm, except maybe the options menu (but, actually we all want to retire the menu bar).
So maybe another look would do the trick, distinguish SNAP from the other toggles?

In Shade we could bend the separator line at the top of the mixer and create space for Snap (and maybe future toggles)
Screenshot_2021-09-22_01-42-51

**Comment by Be-ing** [comment]
There is a merge conflict.

File: src/engine/controls/cuecontrol.cpp [hunk 4]
```diff
@@ -854,13 +856,13 @@ void CueControl::hotcueSet(HotcueControl* pControl, double value, HotcueSetMode
...
     // since it's not necessarily where we currently are. TODO(XXX) is this
     // potentially invalid for vinyl control?
     bool playing = m_pPlay->toBool();
-    if (!playing && m_pQuantizeEnabled->toBool()) {
+    if (!playing && m_pSnapEnabled->toBool()) {
         lock.unlock(); // prevent deadlock.
-        // Enginebuffer will quantize more exactly than we can.
+        // Enginebuffer will snap more exactly than we can.
         seekAbs(cueStartPosition);
     }
```
Comment (Be-ing) [inline src/engine/controls/cuecontrol.cpp:867]: Are you sure this block should be changed to snap? This seeks the deck which I think should solely be done by quantize.

File: src/engine/controls/cuecontrol.cpp [hunk 5]
```diff
@@ -1345,10 +1347,10 @@ void CueControl::cueCDJ(double value) {
...
 
-            // If quantize is enabled, jump to the cue point since it's not
+            // If snap is enabled, jump to the cue point since it's not
             // necessarily where we currently are
-            if (m_pQuantizeEnabled->toBool()) {
-                // Enginebuffer will quantize more exactly than we can.
+            if (m_pSnapEnabled->toBool()) {
+                // Enginebuffer will snap more exactly than we can.
                 seekAbs(mainCuePosition);
             }
```
Comment (Be-ing) [inline src/engine/controls/cuecontrol.cpp:1355]: Same here, I think this should stay using quantize.

File: src/engine/controls/cuecontrol.cpp [hunk 6]
```diff
@@ -1437,10 +1439,10 @@ void CueControl::cuePlay(double value) {
...
             m_pPlay->set(0.0);
-            // If quantize is enabled, jump to the cue point since it's not
+            // If snap is enabled, jump to the cue point since it's not
             // necessarily where we currently are
-            if (m_pQuantizeEnabled->toBool()) {
-                // Enginebuffer will quantize more exactly than we can.
+            if (m_pSnapEnabled->toBool()) {
+                // Enginebuffer will snap more exactly than we can.
                 seekAbs(mainCuePosition);
             }
```
Comment (Be-ing) [inline src/engine/controls/cuecontrol.cpp:1447]: stay with quantize

File: src/engine/controls/cuecontrol.cpp [hunk 7]
```diff
@@ -2133,8 +2135,8 @@ mixxx::audio::FramePos CueControl::quantizeCuePoint(mixxx::audio::FramePos posit
         return mixxx::audio::kInvalidFramePos;
     }
 
-    // Don't quantize when quantization is disabled.
-    if (!m_pQuantizeEnabled->toBool()) {
+    // Don't snap when snapping is disabled.
```
Comment (Be-ing) [inline src/engine/controls/cuecontrol.cpp:2138]:
```suggestion
    // Do nothing when snap is disabled.
```

File: src/engine/controls/loopingcontrol.h [hunk 8]
```diff
@@ -163,6 +163,7 @@ class LoopingControl : public EngineControl {
     ControlValueAtomic<LoopInfo> m_loopInfo;
     LoopInfo m_oldLoopInfo;
     ControlValueAtomic<mixxx::audio::FramePos> m_currentPosition;
+    ControlObject* m_pSnapEnabled;
```
Comment (Be-ing) [inline src/engine/controls/loopingcontrol.h:166]: Use plain class member or std::unique_ptr. No naked `new`. I personally prefer the simplicity of a class member.

**Comment by Be-ing** [comment]
I agree with @ronso0 that the UI could be improved, but I don't think that should hold back merging this. My main concern is that a few places that do seeking have been switched from quantize to snap, but those should continue to use quantize.

Thanks again for working on this. Now that I have a controller with an easily accessible quantize button, I have been using quantize more and this issue has been bugging me frequently.

**Comment by ninomp** [comment]
Merge conflict is resolved. I will try to address review comments tomorrow.

**Review by daschuer — CHANGES_REQUESTED** [review]
Unfortunatly, there is an open discussion that kind of blocks this PR
https://mixxx.zulipchat.com/#narrow/stream/247619-UI-.26.20UX.20design/topic/Decouple.20snap.20from.20quantize

As far as I can see it, we have different expectations, that are sourced from different strategies personal mixing styles, hird party tools and controller layouts. 

Before we continue here we need to check if we either hit the sweet point that fits all or if we have at least the options to make it fit. 

@ninomp Would you mind to describe the solution implemented here in the PR description? 
It is hard to tell from the code where the quantize button is relevant and where the snap button. 
How does it interfere with our moving Main cue? Maybe a table/list of relevant cases would be nice to have an overview.

**Comment by Be-ing** [comment]
@daschuer *please* do *not* leave change requested reviews without specific *changes requested*. You have not raised any specific concerns, only vague, unactionable objections.

**Comment by daschuer** [comment]
@Be-ing 

> Please let's move this discussion to Zulip then. Launchpad is clumsy for discussions.

It was your proposal to move the discussion to Zulip. The Zulip thread contains various specific concerns and proposals. 
It dos not make sense to repeat them here. 

If I skim to the Zulip thread, it does not read as if we reach a conses. But I may misread it. 

My specific request in this review is to add a detailed description about the future behavior this PR initial post.
This gives the chance to verify, if it matches the consensus that may have been reached.
It is also a good source to derive a changelog entry and update the manual.

**Comment by Be-ing** [comment]
Asking for a more detailed description is fine and appropriate. But that is not grounds for a blocking review.

**Review by daschuer — CHANGES_REQUESTED** [review]
Unfortunately the snapping in the fly behavior of the main cue and intro/outro cues does not match the concept of using a single snap button to store cue points. 
I am afraid we need to go back to the cutting table and think about our use cases.
I think it is a welcome feature that the cue points are following the beat grid. 
Maybe we can design quantize/snap around this, instead of copy exactly the Traktor model.

**Comment by Be-ing** [comment]
There are some more merge conflicts now. I would like to get this merged for 2.4. @ninomp can you address @daschuer's review comments above?

**Comment by ninomp** [comment]
Hi @Be-ing! Sorry, I was rather busy with my day work lately. I will come back and finish this soon(ish). Don't worry, I haven't forgotten about this. This is on my to-do list. :slightly_smiling_face:

**Comment by daschuer** [comment]
Thank you for comming back, no need to feel sorry.

Before touching the code, we need to verify if it is a good fit as proposed after all.

I think this issue is sourced by the market situation where the three market leader have three different concepts. 

This PR changes the behavior from the Serato way to the Traktor way. I like the Serato way but can live with this. In terms of "DJ your way" I would be happy to keep an option for the Serato way. But I will not insist.

The related issues are:
* Preferences options should be the last resort
* If we need to find GUI place for another button, the Serato argument for alwing a smaller GUI does not longer apply.
* mapping options 

My original hope was to find a one button solution that allows you workflow, solves you personal anoyance. I can confirm it by the way a solution for this will likely be the best fit for me as well.

Do you see a chance?

**Comment by daschuer** [comment]
Watching myself, there is actually an issue.

I have three use-cases:
1.) Everything quantized 
2.) Set a queue off beat, when the voice starts or such.
3.) Deal with wrong beatgrids.

1. Is obvious.
2. This has an issue: When it is set off beat, and quantize is enabled it produces the same result as if the cue pint has been set with snapping on. So it is mandatory to have quantize off to use the benefit of the exact place, if you forget it = :-(
3. If either the old or the new track has a bad beat grid we meet to disable quantize. For that the hot cue has to be snapped to the Grid already. Issue in the current solution: To forget disabling Quantize after setting the Que.

For 1. Quantize on/Snap don't care.
For 2. Quantize off/Snap off
For 3. Quantize off/Snap on.

Conclusion:

We need to have Quantize accessible during DJing. 

Snap can follow quantize if the DJ is carefully enought and used to it. Users coming from Traktor are probably not ....

Does this summarize the issue?

**Comment by daschuer** [comment]
Two future issues are also related. 

* How to deal with downbeat matching?
* How to deal with Rythms on a  1/16 grid or such.

Will this require another knob or a new set of preferences? It would be nice if this PR is on the straight line towards a solution for this as well. 
To not confuse our user a second time.

**Comment by daschuer** [comment]
Did you consider the text editor auto correct approach: 
* Snap by default 
* Disable snapping after delete and setting the cue again. 

I think that works for me.

**Comment by daschuer** [comment]
So it looks like another button is reasonable if the GUI stays usable ans self explaining in the same way. 

@ronso0 What do you think, is the described issue, of Traktor users worth the work of adding a new button to the GUI? 
Once we have this button, would it be reasonable to hide it again for Serato users? 

What do we need to consider for downbeat matching?

**Comment by Be-ing** [comment]
> I think this issue is sourced by the market situation where the three market leader have three different concepts.

No, the source of the problem is that Mixxx's current design is cumbersome to use and error prone because it is easy to have the timing of playback modified when this is not desired.

> What do we need to consider for downbeat matching?

Please stop demanding to expand the scope of PRs.

**Comment by Be-ing** [comment]
> My original hope was to find a one button solution that allows you workflow

This does not work as explained years ago at the beginning of https://bugs.launchpad.net/mixxx/+bug/1731779

**Comment by ronso0** [comment]
> @ronso0 What do you think, is the described issue, of Traktor users worth the work of adding a new button to the GUI?
Once we have this button, would it be reasonable to hide it again for Serato users?

I didn't bother much with snap and quantize, and I didn't wrap my head around the alternative modes tbh.
Use cases 1, 2 and 3 you described are covered with this, aren't they? Until now DJs had to pay attention to per-deck quantize anyway if they want to start off beat, or when the beat grid is wrong.
At a glance, this PR is a good step forward. I not into this topic to give any feedback on the on-the-fly snapping related to beat grid analysis, sorry.

**Comment by phil-flip** [comment]
Any chance that this gets to be merged. this is one of the issues I am having with quantize.

**Comment by daschuer** [comment]
> Any chance that this gets to be merged. this is one of the issues I am having with quantize.

Unfortunately not. This has some conceptual issues. You may help to do the next step. What is your issue exactly? Please describe what happens and what should happen instead. Did you look out for a different controls? Where did you expect them?

**Comment by phil-flip** [comment]
> > Any chance that this gets to be merged. this is one of the issues I am having with quantize.
> 
> Unfortunately not. This has some conceptual issues. You may help to do the next step. What is your issue exactly? Please describe what happens and what should happen instead. Did you look out for a different controls? Where did you expect them?

I am coming from Rekordbox, and I was looking to have quantize not affect cue and play.
[Rekordbox has settings for that](https://wearecrossfader.co.uk/blog/rekordbox-quantize-settings-how-its-done/), so maybe it can be handled in a similar way, for existing users, while still being able to adjust want to quantize?

**Comment by daschuer** [comment]
The RecordBox solution looks reasonable to me. 
Will you change the configuration from the desired setting one and then?

**Comment by phil-flip** [comment]
Sorry I don't quite understand your question ^^'
Could you rephrase it, please?

**Comment by daschuer** [comment]
In Rekordbox, do you use the same settings all the time or do you change them for different DJing situations. 
Can you post a screenshot of them? 

I can imagine that the main cue and play should be quantized unconditionaly when sync is enabled. Can you confirm?

**Comment by phil-flip** [comment]
Sorry for the late reply.
- I don't actively use Rekordbox, because it doesn't run on Linux. I never did finger drumming, or hot-cues (I'm a beginner). But the beat never snaps on the main clue and only snaps, on play with Sync enabled. (Quantize doesn't matter here)

- When Sync is enabled _with_ quantize? Then this might be a bug, because I never turn on Sync, and it still snaps to the nearest beat. If that is the case then I totally understand your reasoning here, but _only_ quantize shouldn't snap on cue and play.

The only thing that I don't like about Rekordbox: when setting the main cue with Quantize enabled, it always snaps to the nearest beat and the settings don't account for it.

**Comment by phil-flip** [comment]
Any updates on this?

**Comment by daschuer** [comment]
Not yet, we need a plan for the change and a contributor that picks it up. You may file an issue, that describes a low impact change, that would allow your use case.

**Review by Hindikw — APPROVED** [review]

**Comment by PatrickJosh** [comment]
Hi, just to give my opinion on this as I have recently used rekordbox. There, the quantize option does indeed not snap on cue and play. It does, however, make loops and setting cue points snap on beats. I think that this is a reasonable approach.

It is especially annoying to improve your beatmatching skills while simultaneously wanting to be able to set loops with quantize enabled in Mixxx. If I want to improve my beatmatching skills, I now always disable quantize and re-enable it quickly for setting loops because otherwise, the started track automatically snaps to the other one.

Maybe a solution would be to include a similar settings page as the one in rekordbox, especially if this PR has no chance of getting merged in the near future?

**Comment by acolombier** [comment]
I will mark this PR as draft as it seems it is not quite ready to be considered. This is a feature I've also missed in Mixxx and it would be great if we could try to either revive this, or explicitly suggest a way out of the current blockers so anybody willing to revive and complete this PR can do so.
